org.opentosca.bus.management.api.osgievent/src/org/opentosca/bus/management/api/osgievent/MBEventHandler.java

```diff
@@ -157,8 +222,518 @@ public void handleEvent(final Event event) {
+        Set<AbstractNodeTemplate> validDeploymentSubgraph = new HashSet<AbstractNodeTemplate>(nodeTemplates);
+        Collection<AbstractNodeTemplate> toRemove = new HashSet<AbstractNodeTemplate>();
+
+        for (AbstractNodeTemplate nodeTemplate : nodeTemplates) {
+            Collection<AbstractRelationshipTemplate> hostingRelations = this.getOutgoingHostedOnRelations(nodeTemplate);
+            if (!hostingRelations.isEmpty()) {
+                // if we have hostedOn relations check if it is valid under the situation and is in the set
+                boolean foundValidHost = false;
+                for (AbstractRelationshipTemplate relationshipTemplate : hostingRelations) {
+                    AbstractNodeTemplate hostingNode = relationshipTemplate.getTarget();
+                    if (this.isValidUnderSituations(hostingNode, nodeIds2situationIds)) {
+                        if (nodeTemplates.contains(hostingNode)) {
+                            foundValidHost = true;
+                        }
+                    }
+                    if (foundValidHost) {
+                        // if atleast one hosting relationship is fulfilled we are happy
+                        break;
+                    }
+                }
+                if (!foundValidHost) {
+                    toRemove.add(nodeTemplate);
+                }
+            }
+        }
+
+        if (toRemove.isEmpty()) {
+            return validDeploymentSubgraph;
+        } else {
+            validDeploymentSubgraph.removeAll(toRemove);
+            return getDeployableSubgraph(validDeploymentSubgraph, nodeIds2situationIds);
+        }
+    }
+
+    private boolean isValidUnderSituations(AbstractNodeTemplate nodeTemplate,
+                                           Map<String, Collection<Long>> nodeIds2situationIds) {
+
+
+
+        // check if the situation of the policy is active
+        Collection<Long> situationIds = null;
+
+        if ((situationIds = nodeIds2situationIds.get(nodeTemplate.getId())) == null) {
+            return true;
+        }
+        boolean isValid = true;
+
+        for (Long sitId : situationIds) {
+            isValid &= this.isSituationActive(sitId);
+        }
+
+        return isValid;
+    }
+
+    private Collection<AbstractRelationshipTemplate> getOutgoingHostedOnRelations(AbstractNodeTemplate nodeTemplate) {
+        return nodeTemplate.getOutgoingRelations().stream().filter(x -> x.getType().equals(Types.hostedOnRelationType))
+                           .collect(Collectors.toList());
+    }
+
+
+
+    private Collection<AbstractPolicy> getPolicies(QName policyType, AbstractNodeTemplate nodeTemplate) {
+        Collection<AbstractPolicy> policies = new ArrayList<AbstractPolicy>();
+        for (AbstractPolicy policy : nodeTemplate.getPolicies()) {
+            if (policy.getType().getId().equals(policyType)) {
+                policies.add(policy);
+            }
+        }
+        return policies;
```

Comment:
could be reformulated as a `filter` operation on a stream.

Reply:
Done